bn-apps/memberships-management/membership-service/src/main/kotlin/com/r3/businessnetworks/membership/flows/bno/SelfIssueMembershipFlow.kt

```diff
@@ -2,60 +2,106 @@ package com.r3.businessnetworks.membership.flows.bno
-import com.r3.businessnetworks.membership.flows.getAttachmentIdForGenericParam
-import com.r3.businessnetworks.membership.flows.isAttachmentRequired
-import com.r3.businessnetworks.membership.flows.member.RequestMembershipFlow
+import com.r3.businessnetworks.membership.flows.bno.support.BusinessNetworkOperatorFlowLogic
 import com.r3.businessnetworks.membership.flows.member.Utils.throwExceptionIfNotBNO
 import com.r3.businessnetworks.membership.states.MembershipContract
 import com.r3.businessnetworks.membership.states.MembershipState
 import com.r3.businessnetworks.membership.states.MembershipStatus
-import net.corda.core.contracts.Command
-import net.corda.core.flows.*
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.flows.FinalityFlow
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.StartableByRPC
+import net.corda.core.identity.Party
 import net.corda.core.transactions.SignedTransaction
 import net.corda.core.transactions.TransactionBuilder
 import net.corda.core.utilities.ProgressTracker
-import java.time.Instant
 
 /**
- * Self Issue a Membership.  This can only be done by a BNO
+ * Self Issue a Membership.  This can only be done by a BNO,
+ * @param bnoMembership bno membership state to be activated
  */
+@Suppress("UNUSED_VARIABLE")
 @StartableByRPC
 @InitiatingFlow(version = 2)
-open class SelfIssueMembershipFlow(val metaData : Any) : FlowLogic<SignedTransaction>() {
+open class SelfIssueMembershipFlow(val bnoMembership: StateAndRef<MembershipState<Any>>) : FlowLogic<SignedTransaction>() {
     companion object {
-        object RequestingMembership : ProgressTracker.Step("Requesting Membership")
+        object ActivatedMembership : ProgressTracker.Step("Membership Activated")
         object ActivatingMembership : ProgressTracker.Step("Activating Membership")
+        object AlreadyActvie : ProgressTracker.Step("Membership is already active")
 
         fun tracker() = ProgressTracker(
-                RequestingMembership,
-                ActivatingMembership
+                ActivatedMembership,
+                ActivatingMembership,
+                AlreadyActvie
         )
     }
 
     override val progressTracker = tracker()
 
     @Suspendable
-    override fun call() : SignedTransaction {
+
+    //The call function will build a transaction which will only
+    // modify the BNO status and then self sign the transaction
+    override fun call(): SignedTransaction {
+        // stop if Node is not a BNO
         throwExceptionIfNotBNO(ourIdentity, serviceHub)
+        val configuration = serviceHub.cordaService(BNOConfigurationService::class.java)
+        val notary = configuration.notaryParty()
+        logger.info("Membership is being activated")
+        progressTracker.currentStep = ActivatingMembership
+
+
+            val txBuilder = TransactionBuilder(notary)
+                .addInputState(bnoMembership)
```

Comment:
Is this flow meant to do membership issuance? If yes - then where is the existing membership coming from?

Reply:
(line 27) SelfIssueMembershipFlow(val bnoMembership: StateAndRef<MembershipState<Any>>) the parameter bnoMembership will reference the existing bno Member to change it's status